lib/goal/container/k8s.ts

```diff
@@ -0,0 +1,437 @@
+     * provided functionality that the main container needs to
+     * function.  The working directory of the first container is set
+     * to [[ContainerProjectHome]], which contains the project upon
+     * which the goal should operate.
+     */
+    containers: Array<Merge<DeepPartial<k8s.V1Container>, GoalContainer>>;
+    /**
+     * Volumes available to mount in containers.
+     */
+    volumes?: Array<Merge<DeepPartial<k8s.V1Volume>, GoalContainerVolume>>;
+}
+
+/**
+ * Function signature for callback that can modify and return the
+ * [[ContainerRegistration]] object.
+ */
+export type K8sContainerSpecCallback =
+    (r: K8sContainerRegistration, p: GitProject, g: Container, e: SdmGoalEvent, c: HandlerContext) => Promise<K8sGoalContainerSpec>;
+
+/**
+ * Additional options for Kubernetes implementation of container goals.
+ */
+export interface K8sContainerRegistration extends ContainerRegistration {
+    /**
+     * Replace generic containers in [[ContainerRegistration]] with
+     * Kubernetes containers.
+     */
+    containers: Array<Merge<DeepPartial<k8s.V1Container>, GoalContainer>>;
+    /**
+     * Replace generic callback in [[ContainerRegistration]] with
+     * Kubernetes-specific callback.
+     */
+    callback?: K8sContainerSpecCallback;
+    /**
+     * Replace generic volumes in [[ContainerRegistration]] with
+     * Kubernetes volumes.
+     */
+    volumes?: Array<Merge<DeepPartial<k8s.V1Volume>, GoalContainerVolume>>;
+}
+
+export const k8sContainerScheduler: ContainerScheduler = (goal, registration: K8sContainerRegistration) => {
+    goal.addFulfillment({
+        goalExecutor: executeK8sJob(goal, registration),
+        name: DefaultGoalNameGenerator.generateName(`container-k8s-${goal.definition.displayName}`),
+        ...registration as ImplementationRegistration,
+    });
+
+    goal.addFulfillmentCallback({
+        goal,
+        callback: k8sFulfillmentCallback(goal, registration),
+    });
+};
+
+/**
+ * Add Kubernetes job scheduling information to SDM goal event data
+ * for use by the [[KubernetesGoalScheduler]].
+ */
+export function k8sFulfillmentCallback(
+    goal: Container,
+    registration: K8sContainerRegistration,
+): (sge: SdmGoalEvent, rc: RepoContext) => Promise<SdmGoalEvent> {
+
+    return async (goalEvent, repoContext) => {
+        const spec: K8sGoalContainerSpec = _.merge({}, { containers: registration.containers, volumes: registration.volumes });
+        if (registration.callback) {
+            const project = await GitCommandGitProject.cloned(repoContext.credentials, repoContext.id);
+            _.merge(spec, await registration.callback(registration, project, goal, goalEvent, repoContext.context));
+        }
+
+        if (!spec.containers || spec.containers.length < 1) {
+            throw new Error("No containers defined in K8sGoalContainerSpec");
+        }
+
+        spec.containers[0].workingDir = ContainerProjectHome;
```

Comment:
Going to back the synk example: how would I prevent in the k8s case that the workingDir gets set?

Reply:
Here I should probably only set it if it is not set.

Reply:
I've updated this to use the one provided, delete the property of the one provided is an empty string, or set it to `"/atm/home"` if it was not set. Good?